### PR TITLE
[data] fix: make mapping cycles deterministic

### LINKED ISSUE #630: Bring determinism to index over length on mapping dataset
### Feature request

Change the way how mapping datasets in `veomni/data/dataset.py` bring randomness to the data fetching of index over the length of the dataset

### Motivation

When we use some dataset (like the DynamicBatchingDataset) without DistributedSampler on a mapping dataset, the fetching of data might cause a get_item() to be called on the mapping dataset with an index greater than the length of the datset. However, in `veomni/data/dataset.py` we can see that for such an index, the randomness is brought by calling "random.shuffle" on the whole indices array, which is not deterministic because there is no seed defined to control the shuffle.

### Your contribution

It can be updated with a PR to update the way the mapping dataset is shuffled when index is over the length

### PR DESCRIPTION
**Pull request by WenzheWang**
### What does this PR do?

Fixes #630 by replacing process-global in-place shuffling in `MappingDataset` with a reproducible global-index mapping.

The first dataset cycle preserves source order. Each later cycle uses a local RNG seeded by `seed + cycle`, so the selected sample depends only on the global index and seed, not on worker access order, prior reads, or global RNG state. `InterleavedMappingDataset` uses the same mapping.

### Checklist Before Starting

- Search for relative PRs/issues and link here: #630; no matching open PR was returned for #630 or deterministic `MappingDataset` before publication.
- PR title follows `[{modules}] {type}: {description}` format.

### Test

```text
python -m pytest -q tests/data/test_mapping_dataset.py
9 passed

python -m pytest -q tests/data/test_dataloader.py \
  -k 'test_build_dataloader_dyn_bsz_sp_filling and mapping'
8 passed, 14 deselected, 16 warnings
```

The warnings are existing torchdata pin-memory warnings in the CPU test environment.

```text
make quality
All checks passed!
627 files already formatted
```

An independent two-pass review was completed after the final task-code change; it found no remaining correctness or integration blocker.

### API and Usage Example

`build_mapping_dataset` and `InterleavedMappingDataset` accept an optional `seed` (default `42`). Existing callers retain the default behavior contract and the trainer already forwards `args.train.seed`.

```python
dataset = MappingDataset(data, transform=transform, seed=42)
```

### Design & Code Changes

- derive overflow-cycle permutations from a local `random.Random(seed + cycle)` instance;
- retain only the most recently used overflow-cycle permutation and allocate it lazily, avoiding a persistent identity index list per DataLoader worker;
- preserve Python negative-index behavior and raise `IndexError` for empty datasets;
- keep cycle-zero ordering stable after overflow reads and ensure every later cycle is a full permutation;
- inherit the `MappingDataset` constructor directly in `InterleavedMappingDataset`;
- test deterministic access order, restart, seed forwarding through the registered builder, global-RNG isolation, empty input, negative indexing, and interleaved datasets.

### Checklist Before Submitting

- Read the Contribute Guide.
- Applied quality checks.
- API documentation is included in the builder docstring; no user workflow changes.
- Added tests under `tests/data`, which is already included by the GPU unit-test workflow.

### CODE REVIEW

File: tests/data/test_mapping_dataset.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    return [dataset[index] for index in range(start, start + len(dataset))]
+
+
+def test_mapping_dataset_overlength_indices_are_deterministic():
+    first = MappingDataset(list(range(8)), seed=17)
+    second = MappingDataset(list(range(8)), seed=17)
+
+    expected = _read_cycle(first, 1)
+    assert expected != list(range(8))
+    assert _read_cycle(second, 1) == expected
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply tests/data/test_mapping_dataset.py:35]: Added the full-permutation invariant on the current head `8ebe2cf`: the test now asserts `sorted(expected) == list(range(8))`, so dropping or duplicating samples within a cycle is caught directly. The focused MappingDataset suite passes (`9 passed`).

File: veomni/data/dataset.py [hunk 2]
```diff
@@ -53,20 +53,45 @@ def build_dataset(dataset_name: str, **kwargs) -> "Dataset":
...
+        if index < len(self.indices):
+            return self.indices[index]
+
+        cycle, index = divmod(index, len(self.indices))
+        if cycle == self._current_cycle:
+            cycle_indices = self._cycle_indices
+        else:
+            cycle_indices = self.indices.copy()
+            random.Random(self.seed + cycle).shuffle(cycle_indices)
+            self._current_cycle = cycle
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/data/dataset.py:91]: Kept a single later-cycle cache intentionally and documented its bound. The production map-style paths do not feed over-length cycle indices during resume: `StatefulDistributedSampler` indexes the base range, while `_MapStyleSamplerWrapper._rank_indices()` also produces indices in `[0, dataset_size)` and `get_item()` refetches those exact indices. A saved buffer can straddle sampler order, but not `MappingDataset` over-length cycles.

Caching a second full permutation would add another O(N) Python-int list per DataLoader worker. Non-monotonic over-length access remains deterministic and correct; it may rebuild a permutation, which the docstring now states explicitly. The eight mapping dataloader combinations pass on the updated head.

File: veomni/data/dataset.py [hunk 3]
```diff
@@ -53,20 +53,45 @@ def build_dataset(dataset_name: str, **kwargs) -> "Dataset":
...
+    """
+
+    def __init__(self, data: "Dataset", transform: Optional[Callable] = None, seed: int = 42):
         self._data = data
         self._transform = transform
+        self.seed = seed
         self.indices = list(range(len(self._data)))
         self.data_len = len(self.indices)
+        self._current_cycle = 0
+        self._cycle_indices = self.indices
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/data/dataset.py:72]: Updated on `8ebe2cf`: both cache fields now start as `None`, the persistent identity `indices` list has been removed, and base-range/negative indices are mapped arithmetically through `data_len`. Later-cycle storage is allocated only on demand. A repository-wide search found no caller of `MappingDataset.indices`; the focused and mapping dataloader tests pass.

File: veomni/data/dataset.py [hunk 4]
```diff
@@ -1450,7 +1474,7 @@ def build_mapping_dataset(
 
     if transform:
         transform = partial(transform, source_name=source_name)
-    return MappingDataset(data=dataset, transform=transform)
+    return MappingDataset(data=dataset, transform=transform, seed=seed)
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/data/dataset.py:1477]: Added a builder-level regression through `build_dataset(dataset_name="mapping", ...)`. It stubs only dataset I/O, constructs the registered mapping builder with two seeds, and verifies that the over-length cycle changes, so dropping the builder-to-`MappingDataset` seed forwarding is now caught. The focused suite passes (`9 passed`).

File: veomni/data/dataset.py [hunk 5]
```diff
@@ -118,14 +143,11 @@ def __iter__(self):
 
 
 class InterleavedMappingDataset(MappingDataset):
-    def __init__(self, data: "Dataset", transform: Optional[Callable] = None):
-        super().__init__(data, transform)
+    def __init__(self, data: "Dataset", transform: Optional[Callable] = None, seed: int = 42):
+        super().__init__(data, transform, seed)
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/data/dataset.py]: Removed the pass-through `InterleavedMappingDataset.__init__`; it now inherits the complete `MappingDataset` constructor and seed signature directly. The interleaved deterministic-mapping regression remains covered and passes.

**Review by Coach257 — APPROVED** [review]
